compiler/rustc_session/src/options.rs

```diff
@@ -58,18 +59,133 @@ macro_rules! hash_substruct {
+        tmod_info_ret!($parse, $opt_expr, $prefix, $opt_name, $name_hash, $user_value);
+    };
+    ($parse:ident, $opt_expr:expr, $prefix:expr, $opt_name:ident, $name_hash:expr, $user_value:expr, [TRACKED], [TMOD]) => {
+        tmod_info_ret!($parse, $opt_expr, $prefix, $opt_name, $name_hash, $user_value);
+    };
+    ($parse:ident, $opt_expr:expr, $prefix:expr, $opt_name:ident, $name_hash:expr, $user_value:expr, [TRACKED_NO_CRATE_HASH], [TMOD]) => {
+        tmod_info_ret!($parse, $opt_expr, $prefix, $opt_name, $name_hash, $user_value);
+    };
+    ($parse:ident, $opt_expr:expr, $prefix:expr, $_opt_name:ident, $name_hash:expr, $user_value:expr, [SUBSTRUCT], []) => {
+        if let Some(v) = $opt_expr.target_modifier_info_by_hash($name_hash, $user_value) {
+            return Some(v);
+        }
+    };
+    ($parse:ident, $opt_expr:expr, $prefix:expr, $_opt_name:ident, $_name_hash:expr, $user_value:expr, [UNTRACKED], []) => {{}};
+    ($parse:ident, $opt_expr:expr, $prefix:expr, $_opt_name:ident, $_name_hash:expr, $user_value:expr, [TRACKED], []) => {{}};
+    ($parse:ident, $opt_expr:expr, $prefix:expr, $_opt_name:ident, $_name_hash:expr, $user_value:expr, [TRACKED_NO_CRATE_HASH], []) => {{}};
+}
+
+macro_rules! gather_tmods_top_level {
+    ($_opt_name:ident, $opt_expr:expr, $mods:expr, $tmod_vals:expr, [SUBSTRUCT], [TMOD]) => {
+        compile_error!("SUBSTRUCT can't be target modifier");
+    };
+    ($opt_name:ident, $opt_expr:expr, $mods:expr, $tmod_vals:expr, [UNTRACKED], [TMOD]) => {
+        compile_error!("Top level option can't be target modifier");
+    };
+    ($opt_name:ident, $opt_expr:expr, $mods:expr, $tmod_vals:expr, [TRACKED], [TMOD]) => {
+        compile_error!("Top level option can't be target modifier");
+    };
+    ($opt_name:ident, $opt_expr:expr, $mods:expr, $tmod_vals:expr, [TRACKED_NO_CRATE_HASH], [TMOD]) => {
+        compile_error!("Top level option can't be target modifier");
+    };
+    ($_opt_name:ident, $opt_expr:expr, $mods:expr, $tmod_vals:expr, [SUBSTRUCT], []) => {
+        $opt_expr.gather_target_modifiers($mods, $tmod_vals);
+    };
+    ($_opt_name:ident, $_opt_expr:expr, $_mods:expr, $_tmod_vals:expr, [UNTRACKED], []) => {{}};
+    ($_opt_name:ident, $_opt_expr:expr, $_mods:expr, $_tmod_vals:expr, [TRACKED], []) => {{}};
+    ($_opt_name:ident, $_opt_expr:expr, $_mods:expr, $_tmod_vals:expr, [TRACKED_NO_CRATE_HASH], []) => {{}};
+}
+
 macro_rules! top_level_options {
     ( $( #[$top_level_attr:meta] )* pub struct Options { $(
         $( #[$attr:meta] )*
-        $opt:ident : $t:ty [$dep_tracking_marker:ident],
+        $opt:ident : $t:ty [$dep_tracking_marker:ident $( $tmod:ident )?],
     )* } ) => (
         #[derive(Clone)]
         $( #[$top_level_attr] )*
         pub struct Options {
             $(
                 $( #[$attr] )*
                 pub $opt: $t
-            ),*
+            ),*,
+            pub target_modifiers: BTreeMap<String, String>,
```

Comment:
I'm increasingly not a fan of this approach to storing these. I wonder if we could generate alongside `Options`, `CodegenOptions` and `UnstableOptions` some enums, that would look something like:

```rust
enum OptionsTargetModifiers {
    CodegenOptions(CodegenOptionsTargetModifiers),
    UnstableOptions(UnstableOptionsTargetModifiers),
}
enum CodegenOptionsTargetModifiers {
    target_feature,
    // etc.
}
enum UnstableOptionsTargetModifiers {
    regparm,
    reg_struct_return,
    // etc.
}
```

We'd generate a variant for each option marked as a target modifier, and instead of storing a `String` as the key of this map, we could store this enum. Instead of using a hash, we'd use the discriminant in this enum. We could also generate methods which just match on all of the variants and give us the user-facing syntax for the flag:

```rust
impl UnstableOptionsTargetModifiers {
    fn name(&self) -> &'static str {
        match self {
            Self::regparm => "-Zregparm",
            Self::reg_struct_return => "-Zreg-struct-return"
        }
    }
}
```

This should be a lot better than the big function full of if statements and hashes. Do you think that would work?